File: python/pdf-simulator/Check/isChecked.py
import numpy as np
import cv2

def zoom_img(img, scale=1.0):
    center_x, center_y = img.shape[1] / 2, img.shape[0] / 2
    width_scaled, height_scaled = img.shape[1] * scale, img.shape[0] * scale
    left_x, right_x = center_x - width_scaled / 2, center_x + width_scaled / 2
    top_y, bottom_y = center_y - height_scaled / 2, center_y + height_scaled / 2
    img_cropped = img[int(top_y):int(bottom_y), int(left_x):int(right_x)]
    return img_cropped


# Determines whether this image file is checked as yes
def isChecked(path):
    THRESHOLD = 100 # Threshhold for stray marks, in pixels
    CROP = 0.85 # Percentage of image to scan
    PIX_RANGE = 100 # Pixel darkness to consider marked

    img = cv2.imread(path, cv2.IMREAD_GRAYSCALE) # Load image as grayscale
    img_cropped = zoom_img(img, CROP)            # Crop to remove any border

    n_black_pix = np.sum(img_cropped < PIX_RANGE)
    if n_black_pix > THRESHOLD:
        return True
    return False

print("Not selected: ", isChecked("Check/no.png"))
print("Select Line: ", isChecked("Check/yes.png"))
print("Check: ", isChecked("Check/check.png"))
print("Tick: ", isChecked("Check/tick.png"))
print("Dot: ", isChecked("Check/dot.png"))

File: oldDocXtract/pdf-manager/Check/isChecked.js
const cv = require('opencv4nodejs');

function cropImg(img, scale=1.0) {
    const center_x = img.shape[1] / 2;
    const center_y = img.shape[0] / 2;
    const width_scaled = img.shape[1] * scale;
    const height_scaled = img.shape[0] * scale;
    const left_x = center_x - width_scaled / 2;
    const right_x = center_x + width_scaled / 2;
    const top_y = center_y - height_scaled / 2;
    const bottom_y = center_y + height_scaled / 2;
    const img_cropped = img.subarray(top_y, bottom_y, left_x, right_x);
    return img_cropped;
  }

  // Determines whether this image file is checked as yes
  function isChecked(path) {
    const THRESHOLD = 100; // Threshhold for stray marks, in pixels
    const CROP = 0.85; // Percentage of image to scan
    const PIX_RANGE = 100; // Pixel darkness to consider marked
  
    const img = cv.imread(path, cv.IMREAD_GRAYSCALE); // Load image as grayscale
    const img_cropped = cropImg(img, CROP); // Crop to remove any border
  
    const mask = img_cropped.threshold(PIX_RANGE, 255, cv.THRESH_BINARY_INV);
    const numPixels = cv.countNonZero(mask);

    if (numPixels > THRESHOLD) {
      return true;
    }
    return false;
}

console.log("Not selected: ", isChecked("Check/no.png"))
console.log("Select Line: ", isChecked("Check/yes.png"))
console.log("Check: ", isChecked("Check/check.png"))
console.log("Tick: ", isChecked("Check/tick.png"))
console.log("Dot: ", isChecked("Check/dot.png"))